MIOpen resnet50 nightly CI refactor

krzysz00 opened the pull request:
**Note**: before this is merged, I'd like to update the Dockerfile used for this testing to install parallel into the container - calling `apt-get` within a testing pipeline is a bad idea and slows things down, but I wasn't sure which Dockerfile to change and didn't want to break existing tests.

 Add initial revision of MIOpen CI runner refactor

- Refactor the Jenkins file to use a matrix of configurations for
forward/backward directions, different layout, different datatypes,
and whether xdlops should be used
  - This will also allow us to run the non-xdlops tests on gfx900 nodes
- Replace some sh commands with Jenkins plugins, mainly build steps
- Create a new shell script that tests a selection of configs in the
MIOpen environment

This set of changes eliminates a lot of redundant copy-pasted bash
scripts and enables extensiblity.

Some breaking fixes: 
- Remove workspace cleanup as it doesn't work in multi-node settings.
  
  We can consider enabling cleaning before builds, but that'll increase
testing time because there'll be 16 recompiles as opposed to one per
node.

- Force the use of multi_gpu nodes.
  
  This change has been made because single GPU nodes take extremely long
to test certain configs (and may lock up - it's unclear).

These changes have been developed at http://10.236.106.97:8080/job/MLLIR-MIopen-Resnet50-refactor-dev/

krzysz00 committed:
Add initial revision of MIOpen runner refactor

- Refactor the Jenkins file to use a matrix of configurations for
forward/backward directions, different layout, different datatypes,
and whether xdlops should be used
- Replace some sh commands with Jenkins plugins, mainly build steps
- Create a new shell script that tests a selection of configs in the
MIOpen environment

This set of changes eliminates a lot of redundant copy-pasted bash
scripts and enables extensiblity.

krzysz00 committed:
Remove workspace cleanup as it doesn't work in multi-node settings.

We can consider enabling cleaning before builds, but that'll increase
testing time because there'll be 16 recompiles as opposed to one per
node.

krzysz00 committed:
Remove shell scripts non needed in new testing infrastructure.

krzysz00 committed:
Force the use of multi_gpu nodes.

This change has been made because single GPU nodes take extremely long
to test certain configs (and may lock up - it's unclear).

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

krzysz00 commented:
@yiqian1 (and @jerryyin ) for feedback

Also, Yi - with your performance work, how do you gather the statistics? I could add a small Perl script that extracts the `stats:` line and pipes it to some common file, but I'm not sure that's helpful for you.

krzysz00 committed:
Remove stale comment

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
+                stages {
+                    stage('Install parallel (temp until dockerfile updated)') {
+                        steps {
+                            sh 'apt-get update && apt-get install -y --no-install-recommends parallel'
```
Since we are using the static docker image `image 'rocm/mlir:miopen_test'`, there's nothing we can do for this stage now. We should address this problem with testing this pipeline in our channel instead, [here](https://hub.docker.com/repository/docker/rocm/mlir). That can be done incrementally in the follow-up PR.

krzysz00 replied:
Makes sense.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
+                        image 'rocm/mlir:miopen_test'
+                        args '--user "$(id -u):$(id -g)" --device=/dev/kfd --device=/dev/dri --group-add video -u 0'
+                        label "${XDLOPS == '1' ? 'gfx908 && multi_gpu' : 'gfx908 && multi_gpu'}"
+                        // The above could have gfx900 || gfx908 in the latter half
```
Have you tried to use gfx900 nodes in the latter half? I'm not sure we have such a lable, and the one I remember is just labeled as `rocm`.

krzysz00 replied:
There is at least one node labelled gfx900, but I haven't tried this yet, since the dev machine didn't compile miopen (and it's gfx900)

jerryyin replied:
> since the dev machine didn't compile miopen

What's the problem with compiling miopen in gfx900?

krzysz00 replied:
It's been fixed - it was a boost issue

whchung commented:
jenkins: ok to test

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
-                         -DLLVM_ENABLE_TERMINFO=OFF
-                    make -j$(nproc) libMLIRMIOpen
-                '''
+        stage("Carry out test") {
```
[Question] I find a number of quite trivial logs for each matrix cells in your jenkins run # 12. Is it possible to remove those?

Another request is if it is possible to not linear-ize the parallel run of different stage cells. If multiple certain matrix cells has a failure in, for example, installation of a dependency. How do you tell which failure is matched to which run? This might have something to do with how jenkins pipeline is setup, maybe follow up with @okakarpa on how he is able to set up the pipeline in a graph manner. Refer to the CI run log [here](http://10.236.106.97:8080/blue/organizations/jenkins/mlir/detail/PR-215/1/pipeline) as an example.

krzysz00 replied:
If you go to the Blue Ocean link, you'll get a more compact view of the situation - http://10.236.106.97:8080/blue/organizations/jenkins/MLLIR-MIopen-Resnet50-refactor-dev/detail/MLLIR-MIopen-Resnet50-refactor-dev/12/pipeline

krzysz00 replied:
It looks like that big horizontal list is just how the stage view plugin works and there's nothing we can do about it

jerryyin replied:
@krzysz00 Awesome. I think blue ocean view is what I'm asking for.

jerryyin replied:
Please rename `Carry out test` to `Resnet50 configs tests`. It looks like the entire stages aggregation are called `Carry out test` now.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
+                    stage('Build MIOpenDriver') {
+                        steps {
+                            dir('MIOpen') {
+                                git branch: 'igemm-bin-xdlops', poll: false, url: 'https://github.com/ROCmSoftwarePlatform/MIOpen.git'
```
You don't need to checkout to this branch anymore. You can leave it as the default develop branch.

krzysz00 replied:
Done

krzysz00 replied:
As shown in [build 16](http://10.236.106.97:8080/blue/organizations/jenkins/MLLIR-MIopen-Resnet50-refactor-dev/detail/MLLIR-MIopen-Resnet50-refactor-dev/16/pipeline/145) , using the `develop` branch causes the build to fail, so I've reverted the change. Please advise on how to fix this further.

jerryyin replied:
It must not be the branch that breaks it. The two branch should be equivalent because the xdlops PR is merged into develop and the clone does pull the latest commit of MIOpen.

jerryyin replied:
What error message points to is that the [post build action](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-lib/CMakeLists.txt#L75-L80) from cmake you used to comment on is broken, in libMLIRMIOpen build.

krzysz00 replied:
Fixed

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
+                    stage('Build libMLIRMIOpen') {
+                        steps {
+                            checkout scm
+                            cmakeBuild generator: 'Unix Makefiles',\
+                               buildDir: 'build',\
+                               sourceDir: 'llvm',\
+                               installation: 'InDocker',\
+                               buildType: 'Release',\
+                               cmakeArgs: '''-DLLVM_ENABLE_PROJECTS="mlir;lld"
```
Excellent! We should do the same thing with other jenkinsfile too. Could you file a ticket for this?

jerryyin replied:
Actually on a second thought, leave the other jenkinsfile as-is for now. It is already a pain to manually maintain so many jenkinsfile and jenkins job. I don't know which is matched to which now. I believe we can refactor those when we consolidate all jobs to a single one.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
+                    }
+                    stage('Environment') {
+                        steps {
+                            echo "Config is ($DTYPE, $LAYOUT, $XDLOPS, dir: $DIRECTION)"
```
xdlops: $XDLOPS

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
 pipeline {
```
@whchung How would you like this job to be? It looks like this one will replace the original `MLLIR-MIopen-Resnet50-nightly` job. Would you prefer this to be continue to be a nightly job?

jerryyin commented on mlir/utils/jenkins/miopen-tests/miopen_validate.sh:
```diff
@@ -0,0 +1,145 @@
+
+function setup_environment() {
+    export MIOPEN_FIND_MODE=1
+    export MIOPEN_DEBUG_CONV_GEMM=0
```
I believe this can be dropped now.

krzysz00 replied:
We'll see what happens now that I've done that

jerryyin replied:
Please add back `export MIOPEN_FIND_MODE=1`

Sorry for not making it clear, I meant to remove only line 99 and keep line 98.

The problem with removing line 98 is that if MIOpen kernel_db has another algorithm that can run the config, it will not respect the `FIND_ONLY_SOLVER` environment variable. 

The symptom, in the CI log [here](http://10.236.106.97:8080/blue/organizations/jenkins/MLLIR-MIopen-Resnet50-refactor-dev/detail/MLLIR-MIopen-Resnet50-refactor-dev/6/pipeline/359), is that it pick  ConvAsm1x1U instead of the MLIR solver to execute our config.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -1,65 +1,101 @@
-    }
-    post {
-        always {
-            cleanWs()
```
How does it cleanup workspace now?

> We can consider enabling cleaning before builds

Why clean before not after builds?

krzysz00 replied:
The workspace is not cleaned up now - I'm not yet sure how to execute cleanup on all nodes. If I find a way to do that, I'll update.

jerryyin replied:
@krzysz00 Ping @okakarpa if you are not sure. He is the domain expert on this.

krzysz00 replied:
Some googling shows that https://plugins.jenkins.io/hudson-wsclean-plugin/ exists - maybe we want to use it?

jerryyin replied:
@okakarpa Could you comment on that plugin @krzysz00 mentioned about in clean up workspace from distributed environment?

jerryyin reviewed (commented):
Overall excellent work! I think only minor adjustment needed. I'm taking advantage of this PR for some further clarifications to stakeholders.

yiqian1 commented:
Great job! I looked at the console output of your latest job. There are 352 MIOpenDriver runs, but only 314 "Verifies OK". I guess some output got dropped somehow?

krzysz00 committed:
Incorporate most review feedback

- Rename outer-level stage
- Remove a bunch of pointless console output
- Clarify the config names because of the values of XDLOPS (used to be
  xdlops/no-xdlops, now it's 1/0)
 - Remove unneeded branch checkout
 - Remove unneeded environment variables

krzysz00 commented:
Wrt the verifies OK, it looks like some output got snipped. Hopefully trimming the build logs will resolve this.

krzysz00 committed:
Fix insufficient deletion of arg causing errors

krzysz00 committed:
Set the branch correctly

krzysz00 committed:
Put back xdlops branch, the build breaks otherwise

krzysz00 committed:
Revert "Put back xdlops branch, the build breaks otherwise"

This reverts commit 29f3853c1570a3bfdf6349d071ada357169d6d0e.

krzysz00 committed:
Make MIOpen build clean to solve some errors

krzysz00 committed:
Add back in a clean in each workspace

krzysz00 committed:
Add back in a clean in each workspace

krzysz00 committed:
Add some makefile stuff to hopefully fix the build

krzysz00 committed:
Reduce log noise

krzysz00 committed:
Enable non-xdlops on gfx900

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -59,10 +59,11 @@ pipeline {
                     -DCMAKE_RULE_MESSAGES=OFF
                     '''
-                            sh 'cd build; make -j $(nproc) libMLIRMIOpen'
+                            sh 'cd build; make -j $(nproc) libMLIRMIOpen all install'
```
Does this fix the build issue in the jenkins job? I wonder why we need this `all` and `install` targets? It used to be that just building `libMLIRMIOpen` is enough.

krzysz00 replied:
It seems to be the case that the `miopen-test` image had an older version of the mlir headers. This means that, when we used the `develop` branch, the build would try to link in old headers and fail.

The alternate fix is to use the `mlir` docker image and install dependencies.

jerryyin replied:
Thanks!

> The alternate fix is to use the mlir docker image and install dependencies

Way to go!

krzysz00 replied:
I haven't actually done the alternate fix, but it's on my TODO list and will likely be the next PR

krzysz00 replied:
I figure we can keep it as is for now and then follow up in a later PR, since fiddling with the containers is a separate issue

jerryyin commented on mlir/utils/jenkins/miopen-tests/miopen_validate.sh:
```diff
@@ -0,0 +1,143 @@
+declare -g DRIVER="./bin/MIOpenDriver"
+
+function usage() {
+    echo "$0: [-d | --direction] DIR [-t | --dtype] [fp16 | fp32] \
```
I think it is worth explaining the 1: fwd 2: bwd 4: wrw

jerryyin commented on mlir/utils/jenkins/miopen-tests/miopen_validate.sh:
```diff
@@ -0,0 +1,143 @@
+}
+
+function create_configs() {
+    TMPFILE=$(mktemp /tmp/test-config-cmds-miopen.XXXXXX)
```
Could there be interference from, say, if this script is used to populate parallel tests in one environment?

krzysz00 replied:
No. `mktemp` specifically ensures the suffix is unique and is (since there's an underlying system call) thread-safe

jerryyin replied:
Thanks for your explanation.

krzysz00 commented:
Followups (checked means we've put together an internal issue for it):

 - [x] Switch to the mlir docker image and use `cmake -P install-deps.cmake --minimal` to install MIOpen's dependencies
 - [x] Look through other Jenkins files to find something we can refactor (some amount of `when triggeredBy` may end up being used to keep some tests nightly)
 - [x] Generate performance data
 - [x] Re-enable some sort of workspace cleanup

jerryyin commented:
> Followups:
> 

Very good points. You don't have to address those in this PR. But do make sure they end up in a ticket where those follow-up items don't get forgotten.

krzysz00 committed:
Improve help text, add -h

krzysz00 committed:
Add back MIOPEN_FIND_MODE

krzysz00 commented:
> 
> 
> > Followups:
> 
> Very good points. You don't have to address those in this PR. But do make sure they end up in a ticket where those follow-up items don't get forgotten.

Tickets have been filed (for everything but the cleanup bit, which I'm currently talking to @okakarpa about

krzysz00 committed:
Add a clean after each test arm. This'll lengthen testing times but prevent issues with stale builds

yiqian1 commented:
The execution time of each test stage ranges from 17s to 53min, probably because we have limited nodes, and waiting time for available nodes becomes longer and longer when the number of stages increases. Maybe you can try to reduce the matrix size? This could be a serious problem if we add this job to PR-triggered CI. Current MLIR-MIOpen-Resnet-nightly takes 20-25 min.

jerryyin commented:
> The execution time of each test stage ranges from 17s to 53min

I don't see a huge variation on *finished* runs. They range from roughly 40 mins to 70 mins.

> and waiting time for available nodes becomes longer and longer when the number of stages increases.
> This could be a serious problem if we add this job to PR-triggered CI.

I interpret this as the amount of job is simply consuming a lot of resources because of the number of regression tests we run. Those machine resources are shared with tensorflow team as well. We don't have full control over how much fair share of our resource can run.

> Maybe you can try to reduce the matrix size?

I don't see a reason to do that. Instead, what we can do is:
 - For PR triggered CI, cherry-pick a mixture of subset in data type and layout. This reduce the coverage but gives us reasonable confidence.
 - For nightly CI, continue to run everything because worst case scenario 70 mins isn't bad at all

krzysz00 committed:
Update case in workspace clean

krzysz00 committed:
Disable gfx900 nodes - tests hang forever on them.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -11,7 +11,7 @@ pipeline {
                         image 'rocm/mlir:miopen_test'
                         args '--user "$(id -u):$(id -g)" --device=/dev/kfd --device=/dev/dri --group-add video -u 0'
-                        label "${XDLOPS == '1' ? 'gfx908 && multi_gpu' : '(gfx908 || gfx900) && multi_gpu'}"
+                        label "${XDLOPS == '1' ? 'gfx908 && multi_gpu' : 'gfx908 && multi_gpu'}"
```
Please follow up with Omkar on why test hang on gfx900 nodes. Is it because of too many consumers to them?

jerryyin reviewed (approved):
Thanks for your work, LGTM
